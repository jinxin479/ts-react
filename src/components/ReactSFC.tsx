import * as React from "react";
export interface TitleProps {
  address?: string;
  hide?:boolean;
};
export interface ButtonState {
    on?:boolean,
    value?:string
}
export class Title extends React.Component<TitleProps,{}> {
         render() {
             const { address, hide } = this.props;
             return !hide?<a href={address}>{address}</a>:null;
         }
       }
export class Button extends React.Component<{}, ButtonState> {
    state={
        on:false,
        value:""
    }
    onClick = ()=>{
        this.setState(prevState=>{
            return{
                on:!prevState.on,
            }
        })
    }
    onChange = (e: React.FormEvent<HTMLInputElement>)=>{
        this.setState({
            value: e.currentTarget.value
        })
    }
    componentDidMount(){
        console.log(1)/* tslint:disable-line */
    }
    render() {
        const { on,value } = this.state;
        return <React.Fragment>
             <button onClick={this.onClick}>{on ? "on" : "off"}</button> 
             <input onChange={this.onChange}/>
             <span className="input-value">{value}</span>
            </React.Fragment>
    }
}      
const ReactSFC :React.SFC<{message:string}> =({message})=>{
    return <div className="container">
        <h1>welcome use jest</h1>
        <Button/>
        <Title address="台湾选举" hide={false}/>
        {message}
        <span>第一个span</span>
        <span>第二个span</span>
        <ul className="ul-wrapper">
          <li>1</li>
          <li>2</li>
          <li>3</li>
        </ul>
      </div>;
}

export default ReactSFC