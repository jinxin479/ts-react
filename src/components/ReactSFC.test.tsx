import * as React from 'react';
import { shallow, mount} from 'enzyme';
import ReactSFC, { Title, Button} from './ReactSFC';
import toJson from "enzyme-to-json";
/* tslint:disable */
// describe 可以用于测试组，在组里写多个test。it 是test的别名
describe('ReactSFC render test', () => {
    const wrapper = shallow(<ReactSFC message="jinxin" />);
    // console.log(wrapper.debug());
    // expect(wrapper.text()).toEqual("jinxin");
    test("should contain 2 span elements",()=>{
        expect(wrapper.find("span").length).toBe(2);
    });
    test("should contain container className", () => {
        expect(wrapper.find(".container").exists()).toBeTruthy();
    });
    test("ul children length", () => {
        expect(wrapper.find("ul").children().length).toBe(3);
    });
    test("ul has class ul-wrapper", () => {
      expect(wrapper.find("ul").hasClass("ul-wrapper")).toBe(true);
    });
    test("h1 text equal ", () => {
        expect(wrapper.find("h1").text()).toEqual("welcome use jest");
    });
    test("App snapshot",()=>{
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    
});
describe("Title",()=>{
    test("Title prop address", () => {
        const wrapper = shallow(<Title address="https://www.baidu.com" />);
        expect(wrapper.instance().props.address).toBe("https://www.baidu.com");
    });
    test("Title prop href", () => {
      const wrapper = shallow(<Title address="https://www.baidu.com" />);
      expect(wrapper.props().href).toBe("https://www.baidu.com");
    });
    test("Title prop hide", () => {
        const wrapper = shallow(<Title hide={true} address="https://www.baidu.com" />);
        expect(wrapper.find("a").length).toBe(0)
    });
    test("set Title hide prop",()=>{
        const wrapper = shallow(<Title  address="https://www.baidu.com" />);
        wrapper.setProps({hide:true});
        expect(wrapper.find("a").length).toBe(0)

    })
    
})

describe("mount app",()=>{
    test("match snapshot",()=>{
        const tree = mount(<ReactSFC message="mount" />)
        expect(toJson(tree)).toMatchSnapshot();
    })
})

describe("simulate event",()=>{
    test("click button",()=>{
        const wrapper = mount(<ReactSFC message="mount" />);
        const btn = wrapper.find("button");
        //  console.log(wrapper.debug());
        expect(btn.text()).toBe("off");
        btn.simulate("click");
        expect(btn.text()).toBe("on");

    })
    test("onChange event",()=>{
        const wrapper = shallow(<Button/>);
        const input = wrapper.find("input");
        expect(wrapper.find(".input-value").text()).toBe("");
        input.simulate('change', { currentTarget: { value: 'Tyler' } });
        expect(wrapper.find("span.input-value").text()).toBe("Tyler");

    })
    test("button lifecycle", () => {
        jest.spyOn(Button.prototype,"componentDidMount");
        const wrapper = shallow(<Button/>);
        console.log(wrapper.instance());
        expect(Button.prototype.componentDidMount.mock.calls.length).toBe(1);

    })
})