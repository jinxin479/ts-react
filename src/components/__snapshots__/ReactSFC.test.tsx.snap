// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactSFC render test App snapshot 1`] = `
<div
  className="container"
>
  <h1>
    welcome use jest
  </h1>
  <Button />
  <Title
    address="台湾选举"
    hide={false}
  />
  jinxin
  <span>
    第一个span
  </span>
  <span>
    第二个span
  </span>
  <ul
    className="ul-wrapper"
  >
    <li>
      1
    </li>
    <li>
      2
    </li>
    <li>
      3
    </li>
  </ul>
</div>
`;

exports[`mount app match snapshot 1`] = `
<ReactSFC
  message="mount"
>
  <div
    className="container"
  >
    <h1>
      welcome use jest
    </h1>
    <Button>
      <button
        onClick={[Function]}
      >
        off
      </button>
      <input
        onChange={[Function]}
      />
      <span
        className="input-value"
      />
    </Button>
    <Title
      address="台湾选举"
      hide={false}
    >
      <a
        href="台湾选举"
      >
        台湾选举
      </a>
    </Title>
    mount
    <span>
      第一个span
    </span>
    <span>
      第二个span
    </span>
    <ul
      className="ul-wrapper"
    >
      <li>
        1
      </li>
      <li>
        2
      </li>
      <li>
        3
      </li>
    </ul>
  </div>
</ReactSFC>
`;
